paasta_tools/setup_prometheus_adapter_config.py

```diff
@@ -185,9 +236,156 @@ def create_instance_uwsgi_scaling_rule(
         "name": {"as": metric_name},
         "seriesQuery": f"uwsgi_worker_busy{{{worker_filter_terms}}}",
         "resources": {"template": "kube_<<.Resource>>"},
-        # our nicely formatted query has enough whitespace that collapsing said
-        # whitespace cuts down the size of the string by a meaningful amount
-        "metricsQuery": re.sub(pattern=r"\s+", repl=" ", string=metrics_query).strip(),
+        "metricsQuery": _minify_promql(metrics_query),
+    }
+
+
+def should_create_cpu_scaling_rule(
+    autoscaling_config: AutoscalingParamsDict,
+) -> Tuple[bool, Optional[str]]:
+    """
+    Determines whether we should configure the prometheus adapter for a given service.
+    Returns a 2-tuple of (should_create, reason_to_skip)
+    """
+    if autoscaling_config["metrics_provider"] in CPU_METRICS_PROVIDERS:
+        if not autoscaling_config.get("use_prometheus", False):
+            return False, "requested cpu autoscaling, but not using Prometheus"
+
+        return True, None
+
+    return False, "did not request cpu autoscaling"
+
+
+def create_instance_cpu_scaling_rule(
+    service: str,
+    instance: str,
+    autoscaling_config: AutoscalingParamsDict,
+    paasta_cluster: str,
+) -> PrometheusAdapterRule:
+    """
+    Creates a Prometheus adapter rule config for a given service instance.
+    """
+    deployment_name = get_kubernetes_app_name(service=service, instance=instance)
+    sanitized_instance_name = sanitise_kubernetes_name(instance)
+    metric_name = f"{deployment_name}-cpu-prom"
+    moving_average_window = autoscaling_config.get(
+        "moving_average_window_seconds", DEFAULT_CPU_AUTOSCALING_MOVING_AVERAGE_WINDOW
+    )
+
+    # this series query is a bit of a hack: we don't use the Prometheus adapter as expected (i.e., very generic rules)
+    # but we still need to give it a query that returns something even though we're not going to use the series/label
+    # templates that are auto-extracted for us. That said: we still need this query to return labels that can be tied
+    # back to k8s objects WITHOUT using label_replace
+    series_query = f"""
+        kube_deployment_labels{{
+            deployment='{deployment_name}',
+            paasta_cluster='{paasta_cluster}',
+            namespace='paasta'
+        }}
+    """
+
+    cpu_usage = f"""
+        sum(
```

Comment:
Is this `sum()` here because we expect only one `container_cpu_usage_seconds_total` per (pod, container)? Should we `avg()` or `max()` or something instead, just in case we end up with two time series from two different targets?

Reply:
we should only have one series except for bounces - but avg probably makes more sense for handling this